src/inline-assembly.md

```diff
@@ -37,515 +50,892 @@ assert_eq!(x, 4 * 6);
+
+```rust,ignore
+// The following snippet has undefined behaviour
+extern "C-unwind" fn panics(){panic!("unwind through asm")}
+# #[cfg(target_arch = "x86_64")] { unsafe{
+core::arch::asm!("call {}", sym panics);
+# }}
+```
+
+
+r[asm.evaluation.register-value]
+The behaviour is undefined upon exiting an asm block unless the stack pointer register and each operand-usable register not mentioned by an `out` , `lateout`, `inout`, or `inlateout` operand has the value the register held upon entry to the asm block.
+
+> [!TARGET-SPECIFIC]
+> In addition to operand-usable registers, certain other registers on a target may require being preserved, or have specific rules regarding the value at exit.
+> On x86 and x86-64 targets:
+> * The Direction flag (`flags.DF`) is clear upon entry and must be clear upon exit
+> * The x87 Stack (that is the `TOP` field of the floating-point status word, and each bit in the floating-point tag word) must be preserved and restored upon exit. If all x87 `st` registers are marked as clobbered, the stack is guaranteed to be empty on entry to the asm block (that is, `TOP` is set to `0x7` and the `ftw` is set to `0xFFFF`).
+
+## Options [asm.options]
+
+r[asm.options.general]
+An options-spec provided in the asm invocation places constraints on the assembly block. 
+
+r[asm.options.att_syntax]
+The `att_syntax` option may be specfied on the x86 and x86_64 target. The program shall not specify the `att_syntax` option on any other target.
+
+> [!TARGET-SPECIFIC]
+> The `att_syntax` option modifies the syntax used to parse the *expanded asm-string* as though the `.att_syntax prefix` directive was issued before parsing the *expanded asm-string*, and modifies the expansion of register operands to include a `%` prefix.
+
+```rust
+# #[cfg(target_arch = "x86_64")] { unsafe{
+let x: i32;
+core::arch::asm!("mov {:e}, %eax", in(reg) 5, out("eax") x, options(att_syntax));
+# }}
+```
+
+r[asm.options.nomem]
+The `nomem` option may be specified. The behaviour is undefined if the assembly block modifies any allocation, disables, freezes, or activates any tag, *synchronizes-with* any other thread of execution or signal handler, and the implementation may assume that the behaviour or outputs of the assembly block does not depend on the contents of any allocation.
+
+
+```rust,ignore
+// The following snippet has undefined behaviour
+static mut FOO: i32 = 5;
+# #[cfg(target_arch = "x86_64")] { unsafe{
+core::arch::asm!("mov dword ptr [{}+rip], 3", sym FOO, options(nomem));
+# }}
+```
+
+```rust
+// The following snippet may have unpredictable results
+static mut FOO: i32 = 5;
+# #[cfg(target_arch = "x86_64")] { unsafe{
+let x: i32;
+core::arch::asm!("mov {:e}, dword ptr [{}+rip]", out(reg) x, sym FOO, options(nomem));
+# }}
+```
+
+r[asm.options.readonly]
+The `readonly` option may be specified. The behaviour is undefined if the assembly block modifies any allocation or activates any tag. 
```

Comment:
`readonly` also isn't allowed to synchronize with anything. For example you can't use `readonly` for a fence instruction.

Reply:
Are `acquire` operations not allowed for `readonly`? That seems odd to me.

Reply:
I'm not 100% sure if it's legal for the compile to eliminate an acquire load if its result is unused. In fact LLVM currently doesn't do this even for relaxed loads.

Reply:
It is, because it can just take the value from the last store that already *happens-before* so it doesn't make any new side effects visible. LLVM is just understably conservative arround optimizing atomics.